### PR TITLE
Common API for transformation commands and transformation layout

### PR DESCRIPTION
**Pull request by alexsu52**
Introduces the common API for transformation commands and transformation layout and implementation in the TF-backend.

### CODE REVIEW

File: nncf/common/graph/transformations/commands.py [hunk 1]
```diff
@@ -11,14 +11,17 @@
...
 from nncf.common.utils.ordered_enum import OrderedEnum
 
 
 class TransformationPriority(OrderedEnum):
     DEFAULT_PRIORITY = 0
+    FP32_TENSOR_STATISTICS_OBSERVATION = 1
     SPARSIFICATION_PRIORITY = 2
     QUANTIZATION_PRIORITY = 11
-    PRUNING_PRIORITY = 1
+    PRUNING_PRIORITY = 3
```
Comment (ljaljushkin) [inline nncf/common/graph/transformations/commands.py:24]: How does it work before with the first priority? We didn't check it in combination with other algorithms, did we?

File: nncf/common/graph/transformations/commands.py [hunk 2]
```diff
@@ -31,17 +34,22 @@ class TargetType(OrderedEnum):
     LAYER = 0
     BEFORE_LAYER = 1
     AFTER_LAYER = 2
-    WEIGHT_OPERATION = 3
+    PRE_LAYER_OPERATION = 3
```
Comment (ljaljushkin) [inline nncf/common/graph/transformations/commands.py:37]: Can't see usage of `PRE_LAYER_OPERATION` and `POST_LAYER_OPERATION`. Is it just reserved?
Does it match to PRE_OP and POST_OP in NNCf for Pytorch?

File: beta/nncf/tensorflow/graph/model_transformer.py [hunk 3]
```diff
@@ -144,7 +144,7 @@ def _get_target_layer_name(self, target_point):
     def _insert(self, target_point, insertion_objects):
         target_layer_name = self._get_target_layer_name(target_point)
 
-        if target_point.type == TargetType.WEIGHT_OPERATION:
+        if target_point.type == TargetType.LAYER_WEIGHT_OPERATION:
```
Comment (AlexKoff88) [inline beta/nncf/tensorflow/graph/model_transformer.py:147]: What is the intent of renaming this? To me, OPERATION_WITH_WEIGHTS sound more meaningful.

File: beta/nncf/tensorflow/graph/transformations/commands.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+from typing import Any, Callable, List, Optional
+
+from nncf.common.graph.transformations.commands import TargetPoint
+from nncf.common.graph.transformations.commands import TargetType
+from nncf.common.graph.transformations.commands import TransformationCommand
+from nncf.common.graph.transformations.commands import TransformationPriority
+from nncf.common.graph.transformations.commands import TransformationType
+
+
+class TFLayerPoint(TargetPoint):
```
Comment (AlexKoff88) [inline beta/nncf/tensorflow/graph/transformations/commands.py:23]: Can we have short descriptions of these classes that can explain the intent to newcomers? As far as I understand this defines spots for insertion of optimization method-specific operations.

File: nncf/common/graph/transformations/commands.py (hunk 1 above)
Comment (alexsu52) [reply nncf/common/graph/transformations/commands.py:24]: I have two concerns which I address by changing PRUNING_PRIORITY from 1 to 3:
- FP32_TENSOR_STATISTICS_OBSERVATION and PRUNING_PRIORITY have the same priority. This can lead to non-deterministic behavior.
- When Pruning and Sparsity algorithms are stacked. My opinion is that the sparsity mask should be calculated for the pruned weights that have already been pruned by the filter pruning algorithm. Otherwise, the sparsity rate for all models in the sparsity algorithm will be incorrect.

This change influences only the interaction of the sparsity and pruning algorithms. I didn't check it in combination. But If you agree with my statement, I propose to prove this change.

File: nncf/common/graph/transformations/commands.py (hunk 2 above)
Comment (alexsu52) [reply nncf/common/graph/transformations/commands.py:37]: Yes, It is reserved for NNCF with Pytorch backend.
NNCF_MODULE_PRE_OP -> PRE_LAYER_OPERATION 
NNCF_MODULE_POST_OP -> POST_LAYER_OPERATION 

The naming is open. Please, provide your variant. Also, I can exclude these types of target points from this PR. @ljaljushkin What do you think?

Comment (ljaljushkin) [reply nncf/common/graph/transformations/commands.py:37]: I am ok with naming. I was a bit confused to see duplication of similar entities in NNCF PyTorch, like InsertionType (for TransformationType) and OperationPriority(for TransformationPriority).

It was made intentionally, the next step is to re-implement PyTorch part with new classes, right?

File: nncf/common/graph/transformations/commands.py (hunk 1 above)
Comment (ljaljushkin) [reply nncf/common/graph/transformations/commands.py:24]: Sounds reasonable. 
It makes sense to fix [test_priority](https://github.com/openvinotoolkit/nncf/blob/develop/tests/test_nncf_network.py#L257) with new behaviour then.

File: beta/nncf/tensorflow/graph/model_transformer.py (hunk 3 above)
Comment (alexsu52) [reply beta/nncf/tensorflow/graph/model_transformer.py]: I agree, thanks. I have renamed it.

File: beta/nncf/tensorflow/graph/transformations/commands.py (hunk 4 above)
Comment (alexsu52) [reply beta/nncf/tensorflow/graph/transformations/commands.py:23]: Done

File: nncf/common/graph/transformations/commands.py (hunk 2 above)
Comment (alexsu52) [reply nncf/common/graph/transformations/commands.py:37]: Yes, you are right.

File: nncf/common/graph/transformations/commands.py (hunk 1 above)
Comment (alexsu52) [reply nncf/common/graph/transformations/commands.py]: I have added tests. Please take a look.

**Review by ljaljushkin — APPROVED** [review]

**Review by vshampor — APPROVED** [review]
